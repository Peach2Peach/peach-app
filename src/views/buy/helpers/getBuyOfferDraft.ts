type GetBuyOfferDraftParams = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
}
export const getBuyOfferDraft = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
}: GetBuyOfferDraftParams): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
})
