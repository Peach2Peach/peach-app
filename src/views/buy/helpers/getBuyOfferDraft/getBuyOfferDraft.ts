type GetBuyOfferDraftArgs = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
  kyc?: boolean
}
export const getBuyOfferDraft = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
  kyc = false,
}: GetBuyOfferDraftArgs): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
  kyc,
})
