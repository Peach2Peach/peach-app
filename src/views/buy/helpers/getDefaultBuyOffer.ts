type DefaultOfferProps = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
  kyc?: boolean
}
export const getDefaultBuyOffer = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
  kyc = false,
}: DefaultOfferProps): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
  kyc,
})
