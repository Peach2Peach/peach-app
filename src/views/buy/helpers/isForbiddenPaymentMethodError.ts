import { isPaymentMethod } from '../../../utils/validation/isPaymentMethod'

export const isForbiddenPaymentMethodError = (
  errorMessage: string | null,
  errorDetails: unknown,
): errorDetails is PaymentMethod[] =>
  errorMessage === 'FORBIDDEN' && Array.isArray(errorDetails) && errorDetails.every(isPaymentMethod)
