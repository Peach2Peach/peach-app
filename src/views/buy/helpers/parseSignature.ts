const signaturePattern = /[a-zA-Z0-9/=+]{88}/u

export const parseSignature = (signature: string) => signature.match(signaturePattern)?.pop() || signature
