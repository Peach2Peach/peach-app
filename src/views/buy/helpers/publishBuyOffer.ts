import { publishPGPPublicKey } from '../../../init/publishPGPPublicKey'
import { saveOffer } from '../../../utils/offer'
import { postBuyOffer } from '../../../utils/peachAPI'

export const publishBuyOffer = async (offerDraft: BuyOfferDraft) => {
  let [result, err] = await postBuyOffer(offerDraft)

  if (err?.error === 'PGP_MISSING') {
    await publishPGPPublicKey()
    const response = await postBuyOffer(offerDraft)
    result = response[0]
    err = response[1]
  }
  if (result) {
    saveOffer({ ...offerDraft, ...result })
    return { offerId: result.id, isOfferPublished: true, errorMessage: null }
  }
  return { isOfferPublished: false, errorMessage: err?.error || 'POST_OFFER_ERROR' }
}
