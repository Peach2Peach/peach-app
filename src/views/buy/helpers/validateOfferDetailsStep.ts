import { getPaymentData, getSelectedPaymentDataIds } from '../../../utils/account'
import { isDefined } from '../../../utils/array/isDefined'
import { hasMopsConfigured } from '../../../utils/offer'
import { isValidPaymentData } from '../../../utils/paymentMethod'

export const validateOfferDetailsStep = (
  offerDraft: BuyOfferDraft,
  preferredPaymentMethods: Settings['preferredPaymentMethods'],
) =>
  !!offerDraft.amount
  && hasMopsConfigured(offerDraft)
  && getSelectedPaymentDataIds(preferredPaymentMethods).map(getPaymentData)
    .filter(isDefined)
    .every(isValidPaymentData)
