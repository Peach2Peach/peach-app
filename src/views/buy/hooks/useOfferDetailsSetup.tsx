import { Dispatch, SetStateAction, useEffect, useState } from 'react'
import { shallow } from 'zustand/shallow'
import { useHeaderSetup, useShowHelp, useToggleBoolean } from '../../../hooks'
import { useSettingsStore } from '../../../store/settingsStore'
import { account, getPaymentData, getSelectedPaymentDataIds } from '../../../utils/account'
import i18n from '../../../utils/i18n'
import { headerIcons } from '../../../utils/layout/headerIcons'
import { hashPaymentData } from '../../../utils/paymentMethod'
import { validateOfferDetailsStep } from '../helpers/validateOfferDetailsStep'

type Props = {
  offerDraft: BuyOfferDraft
  setOfferDraft: Dispatch<SetStateAction<BuyOfferDraft>>
}
export const useOfferDetailsSetup = ({ offerDraft, setOfferDraft }: Props) => {
  const [isEditing, toggleIsEditing] = useToggleBoolean(false)
  const [isStepValid, setIsStepValid] = useState(false)
  const [meansOfPayment, setMeansOfPayment, preferredPaymentMethods] = useSettingsStore(
    (state) => [state.meansOfPayment, state.setMeansOfPayment, state.preferredPaymentMethods],
    shallow,
  )
  const showHelp = useShowHelp('paymentMethods')

  useHeaderSetup({
    title: i18n(isEditing ? 'paymentMethods.edit.title' : 'paymentMethods.title'),
    icons:
      account.paymentData.length !== 0
        ? [
          {
            ...(isEditing ? headerIcons.checkbox : headerIcons.edit),
            onPress: toggleIsEditing,
          },
          { ...headerIcons.help, onPress: showHelp },
        ]
        : [{ ...headerIcons.help, onPress: showHelp }],
  })

  useEffect(() => {
    const paymentData = getSelectedPaymentDataIds(preferredPaymentMethods)
      .map(getPaymentData)
      .reduce((obj, data) => {
        if (!data) return obj
        obj[data.type] = {
          hash: hashPaymentData(data),
          country: data.country,
        }
        return obj
      }, {} as Offer['paymentData'])

    setOfferDraft((prev) => ({
      ...prev,
      meansOfPayment,
      paymentData,
      originalPaymentData: getSelectedPaymentDataIds(preferredPaymentMethods).map(getPaymentData) as PaymentData[],
    }))
  }, [meansOfPayment, preferredPaymentMethods, setOfferDraft])

  useEffect(
    () => setIsStepValid(validateOfferDetailsStep(offerDraft, preferredPaymentMethods)),
    [offerDraft, preferredPaymentMethods, setIsStepValid],
  )

  return { paymentData: account.paymentData, setMeansOfPayment, isEditing, toggleIsEditing, isStepValid }
}
