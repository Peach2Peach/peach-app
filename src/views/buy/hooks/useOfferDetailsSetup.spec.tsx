import { renderHook } from '@testing-library/react-native'
import { act } from 'react-test-renderer'
import { account1, paymentData } from '../../../../tests/unit/data/accountData'
import { settings1 } from '../../../../tests/unit/data/settingsData'
import { defaultAccount, setAccount } from '../../../utils/account'
import { getBuyOfferDraft } from '../helpers/getBuyOfferDraft'
import { useOfferDetailsSetup } from './useOfferDetailsSetup'
import { headerState, NavigationWrapper } from '../../../../tests/unit/helpers/NavigationWrapper'

const showHelpMock = jest.fn()
const useShowHelpMock = jest.fn((..._args) => showHelpMock)
jest.mock('../../../hooks/useShowHelp', () => ({
  useShowHelp: (...args: any) => useShowHelpMock(...args),
}))

const setMeansOfPaymentMock = jest.fn()
const useSettingsStoreMock = jest.fn((selector) =>
  selector({
    preferredPaymentMethods: settings1.preferredPaymentMethods,
    meansOfPayment: settings1.meansOfPayment,
    setMeansOfPayment: setMeansOfPaymentMock,
  }),
)
jest.mock('../../../store/settingsStore', () => ({
  ...jest.requireActual('../../../store/settingsStore'),
  useSettingsStore: (selector: any) => useSettingsStoreMock(selector),
}))

describe('useOfferDetailsSetup', () => {
  const fakeAccount = { ...account1, paymentData }
  const minBuyAmount = 1
  const maxBuyAmount = 10
  const meansOfPayment = {}
  const offerDraft = getBuyOfferDraft({
    minBuyAmount,
    maxBuyAmount,
    meansOfPayment,
  })
  const setOfferDraftMock = jest.fn()
  const initialProps = {
    offerDraft,
    setOfferDraft: setOfferDraftMock,
  }
  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('should return default values', () => {
    setAccount(defaultAccount)
    const { result } = renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })

    expect(result.current.paymentData).toEqual([])
    expect(result.current.isEditing).toEqual(false)
    expect(result.current.setMeansOfPayment).toEqual(setMeansOfPaymentMock)
    expect(result.current.toggleIsEditing).toBeInstanceOf(Function)
    expect(result.current.isStepValid).toEqual(false)
  })
  it('should return default values with existing payment data', () => {
    setAccount(fakeAccount)
    const { result } = renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })

    expect(result.current.paymentData).toEqual(fakeAccount.paymentData)
  })

  it('should add the correct header', () => {
    setAccount(defaultAccount)
    renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })
    expect(headerState.header()).toMatchSnapshot()
  })
  it('should add correct header for account with payment data', () => {
    setAccount(fakeAccount)
    const { result } = renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })
    expect(headerState.header()).toMatchSnapshot()
    expect(headerState.header().props.icons?.[0].onPress).toEqual(result.current.toggleIsEditing)
  })
  it('should add the correct while editing header', () => {
    setAccount(fakeAccount)
    const { result } = renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })

    act(() => {
      result.current.toggleIsEditing()
    })
    expect(result.current.isEditing).toBeTruthy()
    expect(headerState.header()).toMatchSnapshot()
    expect(headerState.header().props.icons?.[0].onPress).toEqual(result.current.toggleIsEditing)
    expect(headerState.header().props.icons?.[1].onPress).toEqual(showHelpMock)
  })
  it('should update offer draft on init', () => {
    setAccount(defaultAccount)
    renderHook(useOfferDetailsSetup, { wrapper: NavigationWrapper, initialProps })

    expect(setOfferDraftMock).toHaveBeenCalled()
  })
})
