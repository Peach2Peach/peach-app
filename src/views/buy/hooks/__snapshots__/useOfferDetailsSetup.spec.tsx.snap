// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useOfferDetailsSetup should add correct header for account with payment data 1`] = `
<Header
  icons={
    [
      {
        "color": "#7D675E",
        "id": "edit3",
        "onPress": [Function],
      },
      {
        "color": "#099DE2",
        "id": "helpCircle",
        "onPress": [MockFunction],
      },
    ]
  }
  title="payment methods"
/>
`;

exports[`useOfferDetailsSetup should add the correct header 1`] = `
<Header
  icons={
    [
      {
        "color": "#099DE2",
        "id": "helpCircle",
        "onPress": [MockFunction],
      },
    ]
  }
  title="payment methods"
/>
`;

exports[`useOfferDetailsSetup should add the correct while editing header 1`] = `
<Header
  icons={
    [
      {
        "color": "#7D675E",
        "id": "checkboxMark",
        "onPress": [Function],
      },
      {
        "color": "#099DE2",
        "id": "helpCircle",
        "onPress": [MockFunction],
      },
    ]
  }
  title="edit payment methods"
/>
`;
